File: Source/LefticeRuntime/Private/SlateCore/VerticalBoxSlot.cpp
// Copyright (c) NextTurn.
// See LICENSE.txt in the project root for more information.

#if WITH_APPLICATION_CORE

#include "LefticeRuntime/Private/Exports.hpp"

#include "LefticeRuntime/Public/Compiler.hpp"

#include "Runtime/Core/Public/CoreMinimal.h"

#include "Runtime/SlateCore/Public/Widgets/SBoxPanel.h"

ENABLE_WARNINGS;

#define TYPE_EXPORT_MANAGED_TYPE_NAME VerticalBoxSlot

TYPE_EXPORT_START

void EXPORT_CALL_CONV SetHorizontalAlignment(SVerticalBox::FSlot& Slot, EHorizontalAlignment HorizontalAlignment)
{
	Slot.HAlign(HorizontalAlignment);
}

void EXPORT_CALL_CONV SetVerticalAlignment(SVerticalBox::FSlot& Slot, EVerticalAlignment VerticalAlignment)
{
	Slot.VAlign(VerticalAlignment);
}

const FMemberSymbol Members[] =
{
	EXPORT_METHOD(SetHorizontalAlignment),
	EXPORT_METHOD(SetVerticalAlignment),
};

TYPE_EXPORT_END

#undef TYPE_EXPORT_MANAGED_TYPE_NAME

RESTORE_WARNINGS;

#endif

File: Managed/NextTurn.UE.Runtime/Slate/VerticalBoxSlot.cs
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

using NextTurn.UE.Annotations;

namespace Unreal.Slate
{
    public class VerticalBoxSlot : Slot
    {
        private unsafe NativeVerticalBoxSlot* Target => (NativeVerticalBoxSlot*)this.pointer;

        public unsafe int HorizontalAlignment
        {
            get => this.Target->HorizontalAlignment;
            set => NativeMethods.SetHorizontalAlignment(this.Target, value);
        }

        public unsafe int VerticalAlignment
        {
            get => this.Target->VerticalAlignment;
            set => NativeMethods.SetVerticalAlignment(this.Target, value);
        }

        private struct NativeVerticalBoxSlot
        {
            public int HorizontalAlignment { get; }

            public int VerticalAlignment { get; }
        }

        private static class NativeMethods
        {
            [Calli]
            public static extern unsafe void SetHorizontalAlignment(NativeVerticalBoxSlot* slot, int horizontalAlignment);

            [Calli]
            public static extern unsafe void SetVerticalAlignment(NativeVerticalBoxSlot* slot, int verticalAlignment);
        }
    }
}
